students/ejackie/session05/mailroom3.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+
+##################################################
+# Author: Jackie Cheung
+# Date: 2018/11/06
+# Version: 0.1
+# Class: Au2018-Py210B
+# Description: Mailroom Part 3 - Improve your mailroom by adding exceptions and comprehensions.
+##################################################
+
+import sys  # imports go at the top of the file
+
+donor_dict = {("William", "Gates", "III"): [653772.32, 12.17],
+              ("Mark", "Zuckerberg"): [1663.23, 4300.87, 10432.0],
+              ("Jeff", "Bezos"): [877.33],
+              ("Paul", "Allen"): [663.23, 43.87, 1.32]
+              }
+
+main_prompt = "\n".join(("Welcome to the Mailroom!",
+                         "Choose an action:",
+                         "1 - Send a Thank You to a single donor",
+                         "2 - Create a Report",
+                         "3 - Send letters to all donors",
+                         "4 - Quit",
+                         ">>> "))
+
+
+def menu_selection(prompt, dispatch_dict):
+    while True:  # loop forever until they quit.
+        response = input(prompt)
+        try:
+            if dispatch_dict[response]() == "exit program":
+                break
+        except KeyError:
+            print("Please choose a correct number!\n")
+
+
+def send_a_thank_you():
+    name = input("Please enter the full name of the donor: ")
+    if name == "list":
+        [print(" ".join(key)) for key in donor_dict]
+        print()
+    elif tuple(name.split()) in donor_dict:
+        print("Adding donation to existing donor")
+        try:
+            amount = float(input("Please enter the amount of donation: "))
+        except ValueError:
+            print("The amount entered is invalid! Please try again later.\n")
+            return None
+        donor_dict[tuple(name.split())].append(amount)
+        print(send_email(name, amount))
+    else:
+        print("Adding new donor entry to the db")
+        try:
+            amount = float(input("Please enter the amount of donation: "))
+        except ValueError:
+            print("The amount entered is invalid! Please try again later.\n")
+            return None
+        donor_dict[tuple(name.split())] = [amount]
+        print(send_email(name, amount))
```

Comment:
Your lines 44-51 and 53-60 are almost identical (except the print message), so it better be refactored by using `dict.setdefault()` or `defaultdict` or restructuring the `if`-`else` statement block...